eval/adjust.rkt

```diff
@@ -66,53 +65,60 @@
+    (define op (car instr))
     (define tail-registers (cdr instr))
-    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers)) ; tail of the current instr
-    (define output (vector-ref vregs n)) ; output of the current instr
-
-    (define intro (vector-ref vprecs-new (- n varc))) ; intro for the current instruction
-    (define ampls (get-ampls op output srcs)) ; ampls for the tail instructions
-
-    (define final-parent-precision
-      (max (+ intro (vector-ref vstart-precs (- n varc))) (*base-tuning-precision*)))
+    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers))
+    (define output (vector-ref vregs n))
+
+    (define max-prec (vector-ref vprecs-max (- n varc))) ; upper precision bound given from parent
+    (define min-prec (vector-ref vprecs-min (- n varc))) ; lower precision bound given from parent
+
+    ; Final precision assignment based on the upper bound
+    (define final-precision
+      (min (max (+ max-prec (vector-ref vstart-precs (- n varc))) (*rival-min-precision*))
+           (*rival-max-precision*)))
+    (vector-set! vprecs-max (- n varc) final-precision)
+
+    ; Early stopping
+    (match (*lower-bound-early-stopping*)
+      [#t
+       (when (>= min-prec (*rival-max-precision*))
+         (*sampling-iteration* (*rival-max-iterations*)))]
+      [#f
+       (when (equal? final-precision (*rival-max-precision*))
+         (*sampling-iteration* (*rival-max-iterations*)))])
+
+    ; Precision propogation for each tail instruction
+    (define ampl-bounds (get-bounds op output srcs)) ; amplification bounds for children instructions
+    (for ([x (in-list tail-registers)]
+          [bound (in-list ampl-bounds)]
+          #:when (>= x varc)) ; when tail register is not a variable
+      (match-define (cons up-bound lo-bound) bound)
 
-    (when (>= final-parent-precision (*rival-max-precision*)) ; Early stopping
-      (*sampling-iteration* (*rival-max-iterations*)))
+      ; The minimal required precision can not be negative
+      (set! lo-bound (max 0 lo-bound))
 
-    ; Final precision assignment
-    (vector-set! vprecs-new (- n varc) (min final-parent-precision (*rival-max-precision*)))
+      ; Upper precision bound propogation
+      (vector-set! vprecs-max
+                   (- x varc)
+                   (max (vector-ref vprecs-max (- x varc)) (+ max-prec up-bound)))
 
-    ; Intro and ampl propogation for each tail instruction
-    (for ([x (in-list tail-registers)]
-          [ampl (in-list ampls)]
-          #:when (>= x varc)) ; when tail register is not a variable
-      ; check whether this op already has a precision that is higher
-      (when (> (+ intro ampl) (vector-ref vprecs-new (- x varc)))
-        (vector-set! vprecs-new (- x varc) (+ intro ampl))))))
+      ; Lower precision bound propogation
+      (vector-set! vprecs-min
+                   (- x varc)
+                   (max (vector-ref vprecs-min (- x varc)) (+ min-prec lo-bound))))))
```

Comment:
If we really need `lo-bound` non-negative I'd put the `(min lo-bound 0)` here instead of using `set!`. The Racket compiler does not like `set!`.

Reply:
Okay, got it